### PR TITLE
Generalize Minting

### PR DESCRIPTION
**Pull request by mattverse**
Adds generalized minting.

### CODE REVIEW

File: x/concentrated-liquidity/lp_test.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package concentrated_liquidity_test
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (s *KeeperTestSuite) TestMint() {
```
Comment (mattverse) [inline x/concentrated-liquidity/lp_test.go:7]: we use this math in the next mile stone.

File: osmoutils/store_helper.go [hunk 2]
```diff
@@ -106,6 +106,7 @@ func noStopFn([]byte) bool {
 func MustSet(storeObj store.KVStore, key []byte, value proto.Message) {
 	bz, err := proto.Marshal(value)
 	if err != nil {
+		fmt.Println(err.Error())
```
Comment (p0mvn) [inline osmoutils/store_helper.go:109]:
```suggestion
```

File: x/concentrated-liquidity/lp.go [hunk 3]
```diff
@@ -8,24 +8,36 @@ import (
 	types "github.com/osmosis-labs/osmosis/v12/x/concentrated-liquidity/types"
 )
 
-func (k Keeper) Mint(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, tokenIn sdk.Int, lowerTick sdk.Int, upperTick sdk.Int) (numShares sdk.Int, err error) {
-	// first check and validate arguments
+func (k Keeper) Mint(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, liquidityIn sdk.Int, lowerTick sdk.Int, upperTick sdk.Int) (amtDenom0, amtDenom1 sdk.Int, err error) {
 	if lowerTick.GTE(types.MaxTick) || lowerTick.LT(types.MinTick) || upperTick.GT(types.MaxTick) {
-		// TODO: come back to errors
-		return sdk.Int{}, fmt.Errorf("validation fail")
+		return sdk.Int{}, sdk.Int{}, fmt.Errorf("validation fail")
```
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:13]: Can we explain why and potentially format the values?

File: x/concentrated-liquidity/math.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package concentrated_liquidity
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+// liquidity0 takes an amount of asset0 in the pool as well as the sqrtpCur and the nextPrice
+// sqrtPriceA is the smaller of sqrtpCur and the nextPrice
+// sqrtPriceB is the larger of sqrtpCur and the nextPrice
+func liquidity0(amount int64, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:8]: Can we leave formula in the comment?

File: x/concentrated-liquidity/math.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	product := sqrtPriceA.Mul(sqrtPriceB)
+	diff := sqrtPriceB.Sub(sqrtPriceA)
+	amt := sdk.NewDec(amount)
+	return amt.Mul(product.Quo(diff))
+}
+
+// liquidity1 takes an amount of asset1 in the pool as well as the sqrtpCur and the nextPrice
+// sqrtPriceA is the smaller of sqrtpCur and the nextPrice
+// sqrtPriceB is the larger of sqrtpCur and the nextPrice
+func liquidity1(amount int64, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:21]: ditto for formula

File: x/concentrated-liquidity/math.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	diff := sqrtPriceB.Sub(sqrtPriceA)
+	amt := sdk.NewDec(amount)
+	return amt.Quo(diff)
+}
+
+// calcAmount0 takes the asset with the smaller liqudity in the pool as well as the sqrtpCur and the nextPrice and calculates the amount of asset 0
+// sqrtPriceA is the smaller of sqrtpCur and the nextPrice
+// sqrtPriceB is the larger of sqrtpCur and the nextPrice
+func calcAmount0Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:33]: formula please

File: x/concentrated-liquidity/math.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	diff := sqrtPriceB.Sub(sqrtPriceA)
+	mult := liq
+	return (mult.Mul(diff)).Quo(sqrtPriceB).Quo(sqrtPriceA)
+}
+
+// calcAmount1 takes the asset with the smaller liqudity in the pool as well as the sqrtpCur and the nextPrice and calculates the amount of asset 1
+// sqrtPriceA is the smaller of sqrtpCur and the nextPrice
+// sqrtPriceB is the larger of sqrtpCur and the nextPrice
+func calcAmount1Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec) sdk.Dec {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math.go:45]: and here

**Review by p0mvn — APPROVED** [review]

File: x/concentrated-liquidity/lp.go [hunk 8]
```diff
@@ -8,24 +8,36 @@ import (
 	types "github.com/osmosis-labs/osmosis/v12/x/concentrated-liquidity/types"
 )
 
-func (k Keeper) Mint(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, tokenIn sdk.Int, lowerTick sdk.Int, upperTick sdk.Int) (numShares sdk.Int, err error) {
-	// first check and validate arguments
+func (k Keeper) Mint(ctx sdk.Context, poolId uint64, owner sdk.AccAddress, liquidityIn sdk.Int, lowerTick sdk.Int, upperTick sdk.Int) (amtDenom0, amtDenom1 sdk.Int, err error) {
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp.go:11]: as discussed in person, an input of LP shares and an output of respective denoms for mint does not make sense to me, but we can roll with it for now

File: x/concentrated-liquidity/lp.go [hunk 9]
```diff
@@ -8,24 +8,36 @@ import (
...
 	}
 
-	k.UpdateTickWithNewLiquidity(ctx, poolId, lowerTick, tokenIn)
-	k.UpdateTickWithNewLiquidity(ctx, poolId, upperTick, tokenIn)
+	// k.UpdateTickWithNewLiquidity(ctx, poolId, lowerTick, liquidityIn)
+	// k.UpdateTickWithNewLiquidity(ctx, poolId, upperTick, liquidityIn)
 
-	// update tick with new liquidity
-	k.updatePositionWithLiquidity(ctx, poolId, owner.String(), lowerTick, upperTick, tokenIn)
+	// k.updatePositionWithLiquidity(ctx, poolId, owner.String(), lowerTick, upperTick, liquidityIn)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp.go:23]: is this going to get reintroduced?

File: x/concentrated-liquidity/lp_test.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	currentSqrtP, ok := sdk.NewIntFromString("5602277097478614198912276234240")
+	s.Require().True(ok)
+	denom0 := "uosmo"
+	denom1 := "usdc"
+
+	s.SetupTest()
+
+	s.App.ConcentratedLiquidityKeeper.CreateNewConcentratedLiquidityPool(s.Ctx, poolId, denom0, denom1, currentSqrtP, currentTick)
+
+	_, _, err := s.App.ConcentratedLiquidityKeeper.Mint(s.Ctx, poolId, s.TestAccs[0], liquidity, lowerTick, upperTick)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/lp_test.go:31]: should we be testing the asset0 and asset1 outputs are numbers we expect?

File: x/concentrated-liquidity/pool_test.go [hunk 11]
```diff
@@ -1,111 +1,105 @@
...
+
+// 	tooFewPoolDenoms := []string{"usdc"}
+// 	err = pool.SetInitialPoolDenoms(tooFewPoolDenoms)
+// 	require.Error(t, err)
+
+// 	sameDenoms := []string{"usdc", "usdc"}
+// 	err = pool.SetInitialPoolDenoms(sameDenoms)
+// 	require.Error(t, err)
+
+// }
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool_test.go]: we should instead test the new orderInitialPoolDenoms

File: x/concentrated-liquidity/export_test.go [hunk 12]
```diff
@@ -1,6 +1,6 @@
 package concentrated_liquidity
 
 // SetInitialPoolDenoms sets the pool denoms of a cl pool
-func (p *Pool) SetInitialPoolDenoms(poolDenoms []string) error {
-	return p.setInitialPoolDenoms(poolDenoms)
+func (p *Pool) SetInitialPoolDenoms(denom0, denom1 string) error {
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/export_test.go]: Both line 3 and 4 should be called `OrderInitialPoolDenoms`

**Review by czarcas7ic — APPROVED** [review]

File: x/concentrated-liquidity/pool_test.go (hunk 11 above)
Comment (mattverse) [reply x/concentrated-liquidity/pool_test.go]: Added the new test case

File: x/concentrated-liquidity/export_test.go (hunk 12 above)
Comment (mattverse) [reply x/concentrated-liquidity/export_test.go]: changed as well

File: x/concentrated-liquidity/lp.go (hunk 9 above)
Comment (mattverse) [reply x/concentrated-liquidity/lp.go:23]: this is commented out bc we do not have code for initialization

File: x/concentrated-liquidity/lp_test.go (hunk 10 above)
Comment (mattverse) [reply x/concentrated-liquidity/lp_test.go:31]: We don't have the math for it as of right now

**Pull request merged**